keps/sig-scheduling/5710-workload-aware-preemption/README.md

```diff
@@ -0,0 +1,1233 @@
+
+
+## Design Details
+
+### Preemption unit
+
+We start with defining what is a unit of preemption. While we can imagine usecases with preemption
+unit being an arbitrary group of pods, in majority of real world usecases this is actually aligned
+with the scheduling unit. In other words, the group of pods that should be preempted together matches
+a group that was initially scheduled together as a gang.
+
+Trying to formalize it, we define `WorkloadPortion` as one of {all pods in a PodGroup replica or
+a single pod}. With that definition both scheduling unit and preemption units can only be
+`WorkloadPortions`.
+
+In the future, we may want to support usecases when a single scheduling unit consists of multiple
+preemption groups, but we leave that usecase as a future extension (it can be addressed when we
+decide to extend Workload API with PodSubGroup concept - for more details see
+[API Design For Gang and Workload-Aware Scheduling]). However, we never expect preemption unit to
+be larger than scheduling unit.
+
+Based on that, we will extend the the existing `GangSchedulingPolicy` as following:
+
+```golang
+// PreemptionMode describes the mode in which a PodGroup can be preempted.
+// +enum
+type PreemptionMode string
+
+<<[UNRESOLVED PremptionMode vs DisruptionMode]>>
+Should we rename it to DisruptionMode to allow reusing it e.g. in the EvictionRequest API?
+<<[/UNRESOLVED]>>
+
+const (
+  // PreemptionModePod means that individual pods can be preempted independently.
+  // It doesn't depend on exact set of Pods currently running in this PodGroup.
+  PreemptionModePod = "Pod"
+  // PreemptionModePodGroup means that the whole PodGroup replica needs to be
+  // preempted together.
+  PreemptionModePodGroup = "PodGroup"
+)
+
+type GangSchedulingPolicy struct {
+    // Existing field(s).
+
+    // PreemptionMode defines the mode in which a given PodGroup can be preempted.
+    // One of Pod, PodGroup.
+    // Defaults to Pod if unset.
+    PreemptionMode *PreemptionMode
+}
+```
+
+### Workload priorities
+
+Prioritizing workload across each other requires answering the question: "What is workload priority?".
+Up until now, only individual pods have assigned priority. But nothing prevents individual pods
+forming a `Workload` or a `PodGroup` from being heterogenuous and having different priorities.
+
+Intuitively, the priority of a `Workload` or `PodGroup` in such case would be the
+"minimum of priorities of pods that belong to it" - any pod with a priority higher than that can
+e.g. preempt our workload. But intuition is not enough here.
+
+As described in user stories above, a simple static priority doesn't seem to be enough. Arguably it is
+not even a single priority because a priority used for scheduling can be different than the priority
+that should be used for preemption. So in the ideal world a workload owner should be able to define:
+
+- priority used for scheduling (potentially also separately for every PodGroup)
```

Comment:
I do worry about deadlocks if we allow different PodGroups to have different priorities. Imagine two workloads with multiple PodGroups where the priorities interlace.

Reply:
1. That's also why I suggest to start with single priority per workload only
2. Different PodGroups are currently scheduled independently anyway. So we already have this problem. Or maybe I misunderstand your concern here?